### PR TITLE
first stab at a common functions library. include argument parsing and a few definitions wired up. closes #932. closes #945

### PR DESCRIPTION
**Pull request by michaelshobbs**
To be clear, this is a *first* pass. Please have a look and beat it up.

Includes options to enable tracing, auto rm containers on `run`, and "quiet" output. The command line args just set env variables and thus can be used for multiple plugins/commands simultaneously.

This also sets DOKKU_QUIET_OUTPUT if we do not have a tty.

### CODE REVIEW

File: tests/unit/client.bats [hunk 1]
```diff
@@ -21,7 +21,7 @@ teardown() {
 }
 
 @test "dokku client (no args should print help)" {
-  run /bin/bash -c "./contrib/dokku_client.sh | head -1 | grep -q 'dokku COMMAND <app>'"
+  run /bin/bash -c "./contrib/dokku_client.sh | head -1 | egrep -q '^Usage: dokku \[.+\] COMMAND <app>.*'"
```
Comment (josegonzalez) [inline tests/unit/client.bats]: Why this change

Comment (michaelshobbs) [inline tests/unit/client.bats]: The help message header changed. This allows for additional `dokku` args without changing the test.

**Comment by josegonzalez** [comment]
Might make sense to have a shortcut for `DOKKU_QUIET_OUTPUT`?

**Comment by michaelshobbs** [comment]
Shortcut? We expose `--quiet` or `-q` to set it. Is that what you mean?

**Comment by josegonzalez** [comment]
No I mean like an `info2_quiet` function. Not sure how useful that would be in practice though.

**Comment by michaelshobbs** [comment]
Ah! Yeah I was going to do something like that. Forgot between yesterday and today. :smile: I'll take a stab at it.

**Comment by alessio** [comment]
Please use just "name() {}" instead of "function name() {}". The use of `function` is the ugliest POSIX-incompatible bashism :)

**Comment by alessio** [comment]
I'd also suggest to prepend the logging functions (infoN, verbose, warn, etc) with a prefix (dokku_log_$funcname? whatever) to expose them as a sort of consistent public interface to external plugins. A namespace-style prefix would also avoid namespace clashes with functions that could be possibly defined by 3rd-party plugins.

Good job though, thanks!

**Comment by josegonzalez** [comment]
This project is a bash project, and all the files reference the bash interpreter at the top, so I don't see any "bashisms" as bad :) Good call on the function prefixing though.

**Comment by michaelshobbs** [comment]
From http://tldp.org/LDP/abs/html/functions.html:

> Wherever there is repetitive code, when a task repeats with only slight variations in procedure, then consider using a function.

> function function_name { 
> command... 
>} 
>
>or
>function_name () { 
>command... 
>} 
>
>This second form will cheer the hearts of C programmers (and is more portable).

As mentioned, this is distinctly a bash project with no desire to be portable to other shells. That being said if there is additional push back, I don't have a strong position on modifying the `function` syntax.

**Comment by michaelshobbs** [comment]
@alessio i just came across this. https://github.com/progrium/bashstyle
@progrium backs you on not using `function` :smile:

I suppose that's additional push back :wink: I'll refactor it.
